transport/src/main/java/io/netty/channel/AbstractChannel.java

```
@@ -616,24 +616,70 @@ public final void close(final ChannelPromise promise) {
-            outboundBuffer.close(e, true);
+
+            final Throwable shutdownCause = cause == null ?
+                    new ChannelOutputShutdownException("Channel output shutdown") :
+                    new ChannelOutputShutdownException("Channel output shutdown", cause);
+            Executor closeExecutor = prepareToClose();
+            if (closeExecutor != null) {
+                closeExecutor.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        try {
+                            // Execute the shutdown.
+                            doShutdownOutput();
+                            promise.setSuccess();
+                        } catch (Throwable err) {
+                            promise.setFailure(err);
+                        } finally {
+                            // Dispatch to the EventLoop
+                            eventLoop().execute(new Runnable() {
+                                @Override
+                                public void run() {
+                                    closeOutboundBufferForShutdown(pipeline, outboundBuffer, shutdownCause);
+                                }
+                            });
+                        }
+                    }
+                });
+            } else {
+                try {
+                    // Execute the shutdown.
+                    doShutdownOutput();
+                    promise.setSuccess();
+                } catch (Throwable err) {
+                    promise.setFailure(err);
+                } finally {
+                    closeOutboundBufferForShutdown(pipeline, outboundBuffer, shutdownCause);
+                }
+            }
+        }
+
+        private void closeOutboundBufferForShutdown(
+                ChannelPipeline pipeline, ChannelOutboundBuffer buffer, Throwable cause) {
+            buffer.failFlushed(cause, false);
+            buffer.close(cause, true);
```

should we surround this code with a try/catch and fire an exception in the pipeline and/or log if something goes wrong?

---

I think it is fine to just let it bubble up as if it throws there is something seriously wrong.

---

If the task is executed on the EventLoop it won't bubble up will it? It will just log a warning ... is that good enough?

---

Thats true... That said I am not super concerned as if close(...) throws we have a bug in our impl so it should never happen. If you feel better I can definitely add a catch tho. WDYT ?

---

If its a programming error on our end and not expected to ever happen then logging a warning should be "good enough".